src/adc.rs

```diff
@@ -449,6 +449,311 @@ impl_adc!(ADC1: adc_unit_t_ADC_UNIT_1);
 #[cfg(not(any(esp32c2, esp32h2, esp32c5, esp32c6, esp32p4)))] // TODO: Check for esp32c5 and esp32p4
 impl_adc!(ADC2: adc_unit_t_ADC_UNIT_2);
 
+#[cfg(all(not(feature = "riscv-ulp-hal"), not(esp_idf_version_major = "4")))]
+pub mod oneshot {
+    use core::borrow::Borrow;
+
+    use esp_idf_sys::*;
+
+    use crate::gpio::ADCPin;
+    use crate::peripheral::Peripheral;
+    use crate::peripheral::PeripheralRef;
+
+    use super::attenuation::adc_atten_t;
+    use super::config::Resolution;
+    use super::Adc;
+
+    pub mod config {
+        use super::adc_atten_t;
+        use super::Resolution;
+
+        #[derive(Debug, Copy, Clone, Default)]
+        pub struct AdcChannelConfig {
+            pub attenuation: adc_atten_t,
+            pub resolution: Resolution,
+            #[cfg(any(esp_idf_comp_esp_adc_cal_enabled, esp_idf_comp_esp_adc_enabled))]
+            pub calibration: bool,
+        }
+    }
+
+    pub struct AdcChannelDriver<'d, T, ADC, M>
```

Comment:
The `ADC` generic parameter is not necessary. You have it anyway as `T::Adc`.

Reply:
`T::Adc` is defined in terms of ADC, also used in the borrow of AdcDriver.

Reply:
I do not understand where you are going. If you mean, that you have to assert - with `where` - that `M: Borrow<AdcDriver<'d, T::Adc>>` - yes, you have to. Is this syntax not working?

Reply:
ok - now I get it.  Syntax works